File: jenjin-core/src/main/java/com/binarymonks/jj/render/nodes/ShapeRenderNode.java
package com.binarymonks.jj.render.nodes;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.physics.box2d.Transform;
import com.badlogic.gdx.utils.Array;
import com.binarymonks.jj.backend.Global;
import com.binarymonks.jj.pools.N;
import com.binarymonks.jj.pools.Re;
import com.binarymonks.jj.specs.render.RenderSpec;


public abstract class ShapeRenderNode<SPEC extends RenderSpec> extends RenderNode<SPEC> {

    protected boolean fill = true;

    public ShapeRenderNode(SPEC renderSpec, boolean fill) {
        super(renderSpec);
        this.fill = fill;
    }

    @Override
    public void render(OrthographicCamera camera) {
        Global.renderWorld.switchToShapes(fill);
        ShapeRenderer renderer = Global.renderWorld.shapeRenderer;
        renderer.setColor(color.get());
        drawShape(camera);
    }

    protected abstract void drawShape(OrthographicCamera camera);

    public static class Circle extends ShapeRenderNode<RenderSpec> {

        public Vector2 offset = N.ew(Vector2.class);
        public float radius;
        Vector2 positionCache = N.ew(Vector2.class);
        Vector3 positionCache3 = N.ew(Vector3.class);

        public Circle(RenderSpec renderSpec, boolean fill) {
            super(renderSpec, fill);
        }

        @Override
        public void dispose() {

        }

        @Override
        protected void drawShape(OrthographicCamera camera) {
            Transform transform = parent.physicsroot.getTransform();
            positionCache.set(offset);
            transform.mul(positionCache);
            positionCache3.set(positionCache.x, positionCache.y, 0);
            camera.project(positionCache3);
            Global.renderWorld.shapeRenderer.circle(positionCache3.x, positionCache3.y, radius * Global.renderWorld.worldToScreenScale);
        }
    }

    public static class ChainLine extends ShapeRenderNode<RenderSpec> {

        Array<Vector2> vertices;
        Array<Vector3> vertex3Cache = new Array<>();

        public ChainLine(RenderSpec renderSpec, Array<Vector2> vertices) {
            super(renderSpec, false);
            this.vertices=vertices;
        }

        @Override
        public void dispose() {

        }

        @Override
        protected void drawShape(OrthographicCamera camera) {
            clearCaches();
            Transform transform = parent.physicsroot.getTransform();
            for (Vector2 vertex : vertices) {
                Vector2 v2c = N.ew(Vector2.class).set(vertex);
                transform.mul(v2c);
                Vector3 v3c = N.ew(Vector3.class).set(v2c.x, v2c.y, 0);
                camera.project(v3c);
                vertex3Cache.add(v3c);
                Re.cycle(v2c);
            }
            for (int i = 1; i < vertex3Cache.size; i++) {
                Global.renderWorld.shapeRenderer.line(vertex3Cache.get(i - 1), vertex3Cache.get(i));
            }
        }

        private void clearCaches() {
            for (Vector3 cacheVec : vertex3Cache) {
                Re.cycle(cacheVec);
            }
            vertex3Cache.clear();
        }
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/render/nodes/ShapeRenderNode.kt
package com.binarymonks.jj.core.render.nodes

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.math.Matrix3
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Transform
import com.badlogic.gdx.utils.Array
import com.binarymonks.jj.core.JJ
import com.binarymonks.jj.core.pools.new
import com.binarymonks.jj.core.pools.recycle
import com.binarymonks.jj.core.pools.recycleItems
import com.binarymonks.jj.core.pools.vec2
import com.binarymonks.jj.core.properties.PropOverride


abstract class ShapeRenderNode(
        priority: Int,
        color: PropOverride<Color>,
        var fill: Boolean = true
) : RenderNode(priority, color) {

    override fun render(camera: OrthographicCamera) {
        JJ.B.renderWorld.switchToShapes(fill)
        JJ.B.renderWorld.shapeRenderer.color = color.get()
        drawShape(camera)
    }

    abstract fun drawShape(camera: OrthographicCamera)
}

class CircleRenderNode(
        priority: Int,
        color: PropOverride<Color>,
        fill: Boolean = true,
        val offsetX: Float,
        val offsetY: Float,
        val radius: Float,
        var segments: Int = 360
) : ShapeRenderNode(priority, color, fill) {

    private var positionCache: Vector2 = new(Vector2::class)

    override fun drawShape(camera: OrthographicCamera) {
        val transform: Transform = myParent().physicsRoot.transform
        positionCache.set(offsetX, offsetY)
        transform.mul(positionCache)
        JJ.B.renderWorld.shapeRenderer.circle(positionCache.x, positionCache.y, radius, segments)
    }

    override fun dispose() {
    }
}

class LineChainRenderNode(
        priority: Int,
        color: PropOverride<Color>,
        fill: Boolean = false,
        internal var scaleX: Float = 1f,
        internal var scaleY: Float = 1f,
        internal var offsetX: Float = 0f,
        internal var offsetY: Float = 0f,
        internal var rotationD: Float = 0f,
        val vertices: Array<Vector2>
) : ShapeRenderNode(priority, color, fill) {

    private var vertCache: Array<Vector2> = Array()

    override fun drawShape(camera: OrthographicCamera) {
        val transform: Transform = myParent().physicsRoot.transform
        val localTransform = new(Matrix3::class)
        localTransform.scale(scaleX,scaleY)
        localTransform.translate(offsetX,offsetY)
        localTransform.rotate(rotationD)
        clearCache()
        vertices.forEach {
            val transformed = vec2().set(it)
            transform.mul(transformed)
            transformed.mul(localTransform)
            vertCache.add(transformed)
        }
        for (i in 1..vertCache.size - 1) {
            JJ.B.renderWorld.shapeRenderer.line(vertCache.get(i - 1), vertCache.get(i))
        }
        recycle(localTransform)
    }

    override fun dispose() {
    }

    private fun clearCache() {
        recycleItems(vertCache)
        vertCache.clear()
    }
}